### PR TITLE
Fix S1854 FP: ref keyword on a local variable

### LINKED ISSUE #2303: Fix S1854 FP: ref assignments in foreach
### Description

When assigning to `ref` local (obtained via ref return, C# 7 feature), Sonarlint sees it as useless assignment.

### Repro steps

Compiling following snippet produces the S1854 warning.

```csharp
public void Foo()
{
  Span<int> span = new[] { 42 };
  int j = 0;

  foreach (ref var e in span)
    e = j--;
}
```
### Expected behavior

Ne warning to show up.

### Actual behavior

S1854 is produced.

### Known workarounds

Add a suppression.

### Related information

* SonarC# 7.11.0.8083
* .NET Core SDK 2.1.503

### LINKED ISSUE #3304: Fix S1854 FP: ref keyword on a local variable
### Description

When a local variable with ref keyword is used, S1854 is raised.

### Repro steps

```
private void Test(int[] values)
{
    ref int value = ref values[0];
    value = default;
}
```

### Expected behavior

No warning.

### Actual behavior

warning S1854: Remove this useless assignment to local variable 'value'.

### Related information

SonarAnalyzer.CSharp.8.6.0.16497
VisualStudio 2019 16.5.4

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #3304
Fixes #2303

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CFG/ShimLayer/ILocalSymbolExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        private static readonly Func<ILocalSymbol, RefKind> RefKindAccessor;
+
+        static ILocalSymbolExtensions()
+        {
+            RefKindAccessor = LightupHelpers.CreateSyntaxPropertyAccessor<ILocalSymbol, RefKind>(typeof(ILocalSymbol), nameof(RefKind));
+        }
+
+        public static RefKind RefKind(this ILocalSymbol symbol)
+        {
+            return RefKindAccessor(symbol);
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/ShimLayer/ILocalSymbolExtensions.cs:37]: Can be an expression bodied method.

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM

**Comment by pavel-mikula-sonarsource** [comment]
This is a draft to avoid merging to 8.22 milestone.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DeadStores.cs [hunk 2]
```diff
@@ -267,11 +267,12 @@ private void ProcessVariableDeclarator(SyntaxNode instruction, HashSet<ISymbol>
...
-                if (declarator.Initializer != null &&
-                    !IsAllowedInitialization(declarator.Initializer) &&
-                    !symbol.IsConst &&
-                    !liveOut.Contains(symbol) &&
-                    !IsUnusedLocal(symbol))
+                if (declarator.Initializer != null
+                    && !IsAllowedInitialization(declarator.Initializer)
+                    && !symbol.IsConst
+                    && symbol.RefKind() == RefKind.None
+                    && !liveOut.Contains(symbol)
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DeadStores.cs:274]: maybe a HashSet check is cheaper than a reflection call?

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/DeadStores.cs:274]: Property read is almost as fast as reflected property read O(1). I wanted to have the symbol conditions next to each other so I'll leave it as it is.

**Pull request merged**